File: trunk/ShadowFrameworkLite20/src/shadow/pipeline/java/SFGL20ListData.java
package shadow.pipeline.java;

import java.util.ArrayList;

import shadow.system.SFArray;
import shadow.system.SFArrayElementException;

public abstract class SFGL20ListData<E> implements SFArray<E>{

	protected ArrayList<E> data=new ArrayList<E>();
	
	protected abstract E generateGenericElement();
	protected abstract void assignValues(E writing,E reading) throws SFArrayElementException;
	
	@Override
	public void eraseElements(int index, int elementsCount) {
		for (int i = 0; i < elementsCount; i++) {
			data.remove(index);
		}
	}
	
	@Override
	public E generateSample() {
		return generateGenericElement();
	}
	
	//Used only by implementation
	public E getValue(int index){
		return data.get(index);
	}
	
	@Override
	public int generateElement() {
		E e=generateGenericElement();
		data.add(e);
		return data.size()-1;
	}
	
	@Override
	public int generateElements(int count) {
		int index=data.size();
		for (int i = 0; i < count; i++) {
			data.add(generateGenericElement());
		}
		return index;
	}
	
	public void getElement(int index, E element) throws SFArrayElementException {
		assignValues(element, data.get(index));
	}
	
	@Override
	public int getElementsCount() {
		return data.size();
	}
	
	public void setElement(int index, E element) throws SFArrayElementException {
		assignValues(data.get(index),element);
	}
	
}

File: trunk/ShadowFrameworkJS/src/shadow/pipeline/openGL20/SFGL20ListData.js

function SFGL20ListData(){
	this.data=new Array();
}

SFGL20ListData.prototype["eraseElements"]=function(index,elementsCount){
	for (var i = 0; i < elementsCount; i++) {
		this.data.remove(index);
	}
};

SFGL20ListData.prototype["getValue"]=function(index){
	return this.data[index];
};

SFGL20ListData.prototype["generateElement"]=function(index){
	var e=this.generateGenericElement();
	this.data.push(e);
	return this.data.length-1;
};
	
SFGL20ListData.prototype["generateElements"]=function(count){
	var index=this.data.length;
	for (var i = 0; i < count; i++) {
		this.data.push(this.generateGenericElement());
	}
	return index;
};
	
SFGL20ListData.prototype["getElement"]=function(index,element){
		this.assignValues(element, this.data[index]);
	};
	
SFGL20ListData.prototype["setElement"]=function(index,element){
		this.assignValues(this.data[index],element);
	};
	
SFGL20ListData.prototype["getElementsCount"]=function(index,element){
		return this.data.length;
	};

SFGL20ListData.prototype["generateSample"]=function(){
		return this.generateGenericElement();
	};